### ISSUE TITLE
Optimize argmax calculation used in _calc_next_token_id using arm SIMD/Neon instructions

### ISSUE THREAD

**Issue by vimalnakum-sima** [sima-neat/llima#255]
## Feature Summary

LLiMa uses argmax in the _calc_next_token_id which uses bf16 comparisons with OpenMP. It can be further optimized uses arm SIMD/Neon instructions.

## Problem Statement

This would benefit all LLM workloads that use argmax.

## Additional Context

# Vectorizing the greedy-decode argmax in `LanguageModel::_calc_next_token_id`

**Status:** proposal, seeking review
**Target:** `sima_lmm/devkit/cpp/language_model.cpp` on `3.0.0-prep`
**Measured:** +1.71% end-to-end decode TPS on Gemma-4-E2B (som96, 16× Cortex-A65)

---

## 1. Summary

Greedy decode picks the next token with an argmax over the logit vector. For
Gemma-4-E2B that is **262,144 bfloat16 values = 512 KB, scanned once per token**.
The current implementation compares `Eigen::bfloat16` values, which widens every
element to `float` and branches. It costs **~4.9 ns/element single-threaded**, so
it is spread across a `#pragma omp parallel` region over *all* cores to finish in
~256 µs.

Two independent changes are proposed:

| | change | end-to-end TPS |
|---|---|---|
| **A** | cap the OpenMP region at 8 threads instead of all 16 | **+0.90%** |
| **B** | A, plus compare an order-preserving `uint16` key with NEON | **+1.71%** |

They are separable. **A is a one-line change and delivers over half the win**, so
it is worth taking even if B is rejected.

The headline is not just that argmax gets faster (256 µs → 91 µs). It is that the
16-thread region was **slowing down the rest of the token**: capping threads makes
the argmax itself *slower* (256 → 290 µs) and still gains 0.90%, because other
host phases speed up (see §5.2).

---

## 2. Current implementation

```cpp
uint32_t LanguageModel::_calc_next_token_id(MLABuffer* buf_ptr) {
    auto* ptr = reinterpret_cast<Eigen::bfloat16*>(buf_ptr->get_virtual_addr());
    Eigen::bfloat16 max_val = ptr[0];
    uint32_t max_index = 0;
    #pragma omp parallel
    {
        Eigen::bfloat16 thread_max_val = ptr[0];
        uint32_t thread_max_index = 0;
        #pragma omp for nowait
        for (uint32_t i = 0; i < _cfg.lm_cfg.token_cfg.vocab_size; ++i) {
            if (ptr[i] > thread_max_val) { ... }
        }
        #pragma omp critical
        { /* per-thread merge, lowest index wins ties */ }
    }
    return max_index;
}
```

The reduction structure is fine — `critical` runs once per thread, not per
element. Two things make it slow:

1. **Per-element cost.** `Eigen::bfloat16::operator>` widens both operands to
   `float`. With a data-dependent branch that the compiler cannot vectorize
   (the loop carries both a max *and* its index), this measures **1287 µs
   single-threaded**, i.e. ~12 cycles/element on an A65.
2. **It hides that cost behind 16 cores.** The scan parallelizes almost linearly
   (1287 µs @1 → 157 µs @16), which is why the cost was never obvious — but those
   16 threads are contending with the rest of the decode pipeline.

---

### 3.1 Order-preserving integer key

bfloat16 is the top 16 bits of an IEEE-754 binary32. For values of the same sign,
the unsigned bit patterns already order like the numbers (ascending for
positives, descending for negatives). Mapping both halves onto one ascending
unsigned range takes one select:

```
key(u) = (u & 0x8000) ? ~u : (u | 0x8000)
```

- positives `0x0000..0x7FFF` → `0x8000..0xFFFF`, order preserved
- negatives `0x8000..0xFFFF` → `0x7FFF..0x0000`, order reversed (correct: a
  larger magnitude negative is a smaller number)

so `key(a) > key(b)` iff `float(a) > float(b)`, with the two exceptions in §4.
No widening, no branches, and **8 lanes per `vmaxq_u16`**.

The key alone is *not* a win — as scalar code it is **slower** than the current
version (1515 µs vs 1287 µs, because computing the key adds work per element). It
only pays once vectorized. Worth knowing before anyone "simplifies" it back to a
scalar loop.

### 3.2 Two-phase scan, so the hot loop tracks no index

An argmax that carries an index cannot be vectorized cleanly (you need a
per-lane index vector, and 262,144 > `UINT16_MAX` so the lanes would have to
widen). Instead split it:

- **Phase 1** — for each 8192-element chunk, compute the max *key* only: four
  `vmaxq_u16` accumulators, no indices, no branches, then one `vmaxvq_u16`.
  Track which chunk holds the global max, updating on **strict `>`** so the
  *first* such chunk wins.
- **Phase 2** — rescan that one chunk (≤8192 elements) for the first index whose
  key equals the global max.

Phase 1 is the only pass over 512 KB. Phase 2 touches ≤16 KB, already warm.

### 3.3 Thread count

`num_threads(std::min(8, omp_get_max_threads()))`. See §5.2 — 8 is measured
optimal on this SoC and 12/16 regress. `embedding_offload.cpp:103` already uses
`num_threads(8)`, so this matches existing practice in the codebase rather than
introducing a new convention.

---

## 4. Behavioural differences

Both are reachable only from logits that already indicate a broken model, but
they are real and reviewers should decide explicitly:

| input | current | proposed |
|---|---|---|
| **NaN present** | skipped (`v > max` is false for NaN), returns the max non-NaN | a NaN **wins** — positive NaN maps to the top of the key range |
| **max is `-0.0`, `+0.0` also present** | compare equal, lowest index wins | `+0.0` sorts above `-0.0`, so the `+0.0` index wins |

`+Inf` is unaffected: `0x7F80 | 0x8000 = 0xFF80` sorts above every finite
positive, matching float semantics.

Arguably surfacing a NaN is preferable to silently returning a plausible token,
but it *is* a change. If the NaN behaviour must be preserved exactly, a
`vceqq_u16` NaN mask in phase 1 would cost roughly one extra instruction per
vector — we have not implemented or measured that.

**Tie-break is preserved.** The current code returns the lowest index among
equal maxima; the strict `>` on the chunk pick plus the forward scan in phase 2
reproduce that. Verified empirically against the current implementation on 32
random 262,144-element buffers: zero mismatches.

---

## 5. Measurements

Board: som96, 16× Cortex-A65 @ `-mcpu=cortex-a65 -march=armv8.2-a+crypto -O3
-fopenmp` (the project's own flags).

### 5.1 Isolated microbenchmark

`/workspace/argmax-bench.cpp`. 32 separate 512 KB buffers rotated round-robin
over 16 MB, so **every scan starts cold** — matching llima, where the logits
arrive by DMA and are `invalidate_cache()`d before the scan. 200 reps.

| variant | µs/scan |
|---|---|
| current, 16 threads | 156.7 |
| current, 8 / 4 / 2 threads | 285.6 / 430.4 / 648.7 |
| current, 1 thread | 1286.6 |
| integer key, scalar, 1 thread | **1515.3** ← slower; the key only pays vectorized |
| **NEON, 1 thread** | **165.4** ← one core ≈ all 16 cores of the current code |
| NEON, 2 / 4 threads | 110.9 / 67.1 |
| **NEON, 8 threads** | **53.9** |
| NEON, 12 / 16 threads | 80.8 / 77.2 ← regresses, bandwidth-bound |

### 5.2 End-to-end llima decode

`sima_lmm_text_generation_test`, `Gemma-4-E2B-it-TextOnly-GPTQ-a16w4`, prompt
"Why is the sky blue?", 256 tokens. Phase timings are means over **tokens 3..258**
(tokens 1–2 are prefill at ~81 ms and would swamp the mean). Arms interleaved
run-by-run so thermal or memory drift cannot favour either; each arm's `.so` is
md5-gated before *and* after every run, aborting rather than reporting if the
wrong binary is resident. Both arms are built from one tree behind a
`#define`, so the class layout — and therefore the test binary — is identical and
the delta is attributable to this change alone.

| arm | argmax µs | `bind` µs | total µs | TPS avg | vs A |
|---|---|---|---|---|---|
| **A** current, 16 threads (n=5) | 255.6 | **492.3** | 27,975 | 35.876 (sd .036) | — |
| **C** current, `OMP_NUM_THREADS=8` (n=3) | 289.7 | **390.7** | 27,786 | 36.200 (sd .061) | **+0.90%** |
| **B** NEON key, 8 threads (n=8) | **90.8** | 384.9 | 27,579 | 36.489 (sd .063) | **+1.71%** |

Welch t: A→C 8.4, C→B 7.0, A→B 22.4. All three significant; TPS ranges do not
overlap between arms.

**The contention effect (A→C) is the surprising half.** Capping threads makes the
argmax phase itself *worse* (255.6 → 289.7 µs, exactly as §5.1 predicts) and the
token still gets faster, because `bind` — unrelated host work that runs *before*
argmax — drops **492 → 391 µs**, and `queue` drops ~119 µs. The leading
explanation is GNU OpenMP's default spin-wait: 16 workers busy-wait after the
region and steal cycles from the main thread and from MLA-RT's submit/reap.

⚠ **This explanation is not yet confirmed.** The direct test —
`OMP_WAIT_POLICY=passive` with 16 threads, which should recover most of the A→C
gain if spin-waiting is the mechanism — is still outstanding. Both som96 and
dvt60 were rebooted with an updated MLA driver on 2026-09-23, and the installed
`libMLArt-K.so` (`e1a90f37…`) no longer matches it:

```
mla 5000000.mla: mla_op_job_async: EINVAL sqe->req_len=104 sizeof(req)=88
```

llima gets through model load and prefill, then every job is rejected and decode
produces zero tokens with no error in the llima log — so **rebuild libMLArt
against the new UAPI before attempting any llima run**, on either board. If the
mechanism turns out to be something other than spin-waiting, the *measurements*
above still stand; only the explanation of the A→C half would change.

**Correctness:** the 258 generated tokens are byte-identical across all 16 runs
of all three arms and to the untouched baseline (md5
`ee5d8d76810aa6938bad80a781907c86`).

### 5.3 Where this sits in the token budget

97.2% of the token is MLA time (`queue` ≈ 27,130 µs of 27,975 µs). Total
host-side work is ~760 µs. Argmax was **a third of all host work** and is the
only host phase that responded to optimization — we separately implemented and
measured four `make_bindings` micro-optimizations (memoization, override-map
merge walk, offset caching) and every one came out at ~0, because that code is
memory-latency-bound over scattered per-model structures rather than
compute-bound. Happy to share those negative results if useful.

---

## 6. Proposed patch

Instrumentation and the A/B `#define` removed; this is the production form.

```cpp
#if defined(__aarch64__)
#include <arm_neon.h>
#endif
#include <omp.h>
#include <algorithm>

namespace {

// Order-preserving unsigned key for a bfloat16 bit pattern: see design note.
// key(a) > key(b) iff float(a) > float(b), except that NaN sorts to the top and
// -0.0 sorts just below +0.0.
inline uint16_t bf16_order_key(uint16_t bits) {
    return (bits & 0x8000u) ? static_cast<uint16_t>(~bits)
                            : static_cast<uint16_t>(bits | 0x8000u);
}

}  // namespace

uint32_t LanguageModel::_calc_next_token_id(MLABuffer* buf_ptr) {
    const uint32_t vocab = _cfg.lm_cfg.token_cfg.vocab_size;
#if defined(__aarch64__)
    // Phase 1 takes the max of the order-preserving keys per chunk with no index
    // tracking, so it vectorizes to 8 lanes with no branches; phase 2 then finds
    // the index inside the single chunk that achieved the global max. Comparing
    // Eigen::bfloat16 directly widens every element to float and costs ~12
    // cycles each, which needed all 16 cores to stay off the critical path.
    {
        const uint16_t* const p =
            reinterpret_cast<const uint16_t*>(buf_ptr->get_virtual_addr());
        constexpr uint32_t kChunk = 8192;
        const uint32_t nchunks = (vocab + kChunk - 1) / kChunk;

        // Reused across tokens to keep argmax allocation-free on the hot path.
        // NOTE: the raw pointer is taken *outside* the parallel region on
        // purpose -- naming a thread_local inside it would resolve to each
        // worker's own copy and the results would be dropped.
        thread_local std::vector<uint16_t> chunk_max_storage;
        chunk_max_storage.assign(nchunks, 0);
        uint16_t* const chunk_max = chunk_max_storage.data();

        // 8 threads, not all cores: measured optimal here, and a wider region
        // steals cycles from the main thread and MLA-RT for the rest of the
        // token. Matches embedding_offload.cpp's existing num_threads(8).
        const int nthreads = std::min(8, omp_get_max_threads());
        #pragma omp parallel for num_threads(nthreads) schedule(static)
        for (uint32_t c = 0; c < nchunks; ++c) {
            const uint32_t off = c * kChunk;
            const uint32_t n = (off + kChunk <= vocab) ? kChunk : vocab - off;
            const uint16_t* const q = p + off;

            const uint16x8_t signbit = vdupq_n_u16(0x8000);
            uint16x8_t m0 = vdupq_n_u16(0), m1 = m0, m2 = m0, m3 = m0;
            uint32_t i = 0;
            for (; i + 32 <= n; i += 32) {
                const uint16x8_t a = vld1q_u16(q + i);
                const uint16x8_t b = vld1q_u16(q + i + 8);
                const uint16x8_t c2 = vld1q_u16(q + i + 16);
                const uint16x8_t d = vld1q_u16(q + i + 24);
                m0 = vmaxq_u16(m0, vbslq_u16(vtstq_u16(a, signbit),
                                             vmvnq_u16(a), vorrq_u16(a, signbit)));
                m1 = vmaxq_u16(m1, vbslq_u16(vtstq_u16(b, signbit),
                                             vmvnq_u16(b), vorrq_u16(b, signbit)));
                m2 = vmaxq_u16(m2, vbslq_u16(vtstq_u16(c2, signbit),
                                             vmvnq_u16(c2), vorrq_u16(c2, signbit)));
                m3 = vmaxq_u16(m3, vbslq_u16(vtstq_u16(d, signbit),
                                             vmvnq_u16(d), vorrq_u16(d, signbit)));
            }
            uint16_t best =
                vmaxvq_u16(vmaxq_u16(vmaxq_u16(m0, m1), vmaxq_u16(m2, m3)));
            for (; i < n; ++i) {                     // tail: vocab need not be
                const uint16_t k = bf16_order_key(q[i]);   // a multiple of 32
                if (k > best) best = k;
            }
            chunk_max[c] = best;
        }

        uint16_t gmax = 0;
        uint32_t gchunk = 0;
        for (uint32_t c = 0; c < nchunks; ++c) {
            // Strict >: keeps the FIRST chunk holding the max, which is what
            // makes phase 2 return the lowest index overall.
            if (chunk_max[c] > gmax) { gmax = chunk_max[c]; gchunk = c; }
        }
        const uint32_t off = gchunk * kChunk;
        const uint32_t n = (off + kChunk <= vocab) ? kChunk : vocab - off;
        for (uint32_t i = 0; i < n; ++i) {
            if (bf16_order_key(p[off + i]) == gmax) return off + i;
        }
        return off;
    }
#else
    // ... existing Eigen::bfloat16 OpenMP loop, unchanged, as the portable path
#endif
}
```

Notes for review:

- **`vocab` need not be a multiple of 32 or of `kChunk`.** The last chunk is
  short and the in-chunk tail is scalar. This matters in practice: the
  `gemma3-siglip448-a16w4` config declares `vocab_size` **262,208** = 32 full
  chunks + one 64-element chunk, versus 262,144 (exactly 32 chunks) for the model
  we measured. ⚠ **The short-chunk path is reasoned-about, not measured** — every
  number in §5 comes from the 262,144 case. Please exercise a non-multiple vocab
  before merging.
- The `thread_local` buffer avoids a per-token heap allocation. A member on
  `LanguageModel` would work too, but that changes the class layout and forces a
  rebuild of everything constructing one, so `thread_local` seemed the lighter
  option. Your call.
- Scoped in a block so the `#else` path is a drop-in.

---

## 7. What has **not** been tested

Listed so nobody over-reads the result:

1. **One model, one prompt, one board.** Gemma-4-E2B-it-TextOnly-GPTQ-a16w4,
   "Why is the sky blue?", 256 tokens, greedy decode, **som96 only**. Not run on
   dvt60 at all, not with a vision model, not with a different `vocab_size`.
   Reproducing on dvt60 needs setup first: it has no `libsima_lmm_runtime.so`, no
   `sima_lmm/assets` directory, and no Gemma-4 corpus — only
   `gemma3-siglip448-a16w4` and two LoRA variants. A vision model also brings the
   `vm_cfg` branches in the decode loop into play, so its baseline TPS will not be
   comparable to the numbers here.
2. **Speculative decode and the sampling path are untouched and unmeasured.**
   This only changes greedy argmax. The `_need_argmax == false` path (where the
   model emits the token id directly) is unaffected.
3. **`num_threads(8)` is tuned for 16× Cortex-A65.** On a different core count
   the optimum will differ; `std::min(8, omp_get_max_threads())` is a guess
   outside 16 cores. If you would rather derive it, or make it configurable,
   that is a reasonable review outcome.
4. **The spin-wait explanation for the A→C half is unconfirmed** — see §5.2.
5. **NaN/-0.0 semantics change** — §4.

## 8. Open questions for reviewers

1. Take **A only** (one-line thread cap, +0.90%, zero risk) as a separate quick
   change, then evaluate B on its own?
2. Is the NaN behaviour change acceptable, or should it be masked?
3. `num_threads(8)`: hardcode, derive, or expose as configuration?
4. Worth applying the same thread-count scrutiny to `utils.hpp`'s `topk`
   `#pragma omp parallel for`? It parallelizes over *rows*, so it is effectively
   serial for greedy decode (rows = 1) and we did not measure it — but it is
   uncapped, and on a sampling path with many rows it would fan to all 16 cores
   in the same way.
5. **Bigger structural question this raises:** `_need_argmax == false` implies
   some models emit the token id on-device. If the Gemma-4 ELF could do the
   argmax on the MLA, both the remaining ~91 µs *and* the 50 µs
   `invalidate_cache()` of 512 KB would disappear — worth roughly another 0.5%.
   That is a model-compile change, not a runtime one, but it may be the better
   long-term answer.

## 9. Reproducing

- Microbenchmark: `/workspace/argmax-bench.cpp` — self-contained, prints the
  §5.1 table and cross-checks every variant against the current implementation.
  `g++ -mcpu=cortex-a65 -march=armv8.2-a+crypto -O3 -fopenmp argmax-bench.cpp`
- End-to-end: build `sima_lmm_runtime` + `sima_lmm_text_generation_test` from
  `3.0.0-prep` with the patch behind a `#define`, flip it for the two arms so the
  test binary is shared, deploy each arm to its own directory and select with
  `LD_LIBRARY_PATH`, and interleave the runs. Phase timings come from local
  instrumentation (a per-token `PHASES us:` line); TPS comes from the test's own
  `TPS: ... avg=` line — **use `avg`, not `max`**.

**Comment by vimalnakum-sima** [comment]
[argmax-bench.cpp](https://github.com/user-attachments/files/32620549/argmax-bench.cpp)
argmax example implementation and workload measurement
